test/framework/integration_framework/integration_test_framework.cpp

```diff
@@ -133,26 +129,103 @@ class IntegrationTestFramework::CheckerQueue {
     return obj;
   }
 
-  boost::optional<T> try_pop() {
+  std::optional<T> try_pop() {
     std::unique_lock<std::mutex> lock(queue_mutex_);
     if (queue_.empty()) {
       if (not cv_.wait_for(
               lock, timeout_, [this] { return not queue_.empty(); })) {
-        return boost::none;
+        return std::nullopt;
       }
     }
     T obj(std::move(queue_.front()));
     queue_.pop();
     return obj;
   }
 
+  auto size() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    return queue_.size();
+  }
+
  private:
   std::chrono::milliseconds timeout_;
   std::queue<T> queue_;
   std::mutex queue_mutex_;
   std::condition_variable cv_;
 };
 
+#include <unordered_map>
+
+struct IntegrationTestFramework::ResponsesQueues {
+ public:
+  using HashType = shared_model::interface::types::HashType;
+
+ private:
+  /// maximum time of waiting before appearing next transaction response
+  std::chrono::milliseconds tx_response_waiting_time_ms;
+  std::recursive_mutex mtx;
```

Comment:
You don't need recursive mutex here. It is a bad way to lock outside and enter under mutex twice

Reply:
reworked